### ISSUE TITLE
Different error message vanilla vs CS (again)

### ISSUE THREAD

**Issue by sorawee** [racket/racket#3325]
**What version of Racket are you using?**
7.8.0.6--2020-07-31 [cs]

**What program did you run?**

```
#lang racket

(define x y)
(define y 1)
```

**What should have happened?**

```
y: undefined;
 cannot reference an identifier before its definition
```

in DrRacket, and 
```
y: undefined;
 cannot reference an identifier before its definition
  in module: "/path/to/test.rkt"
```

in Racket.

**If you got an error message, please include it here.**

```
../../../racket/racket/share/pkgs/drracket/drracket/private/stack-checkpoint.rkt:115:30: define-values: skipped variable definition;
 cannot continue without defining variable
  variable: x
  in module: "/path/to/test.rkt"
```

in DrRacket when "Enforce constant definitions (enables some inlining)" is off.

```
../../../racket/racket/share/pkgs/drracket/drracket/private/stack-checkpoint.rkt:115:30: y: undefined;
 cannot reference an identifier before its definition
```

in DrRacket when "Enforce constant definitions (enables some inlining)" is on.

```
y: undefined;
 cannot reference an identifier before its definition
  module: "/path/to/test.rkt"
```

in Racket.

**Please include any other relevant details**

1. Notice that `git/racket/racket/share/pkgs/drracket/drracket/private/stack-checkpoint.rkt:115:30` doesn't exist in DrRacket BC.
2. In DrRacket BC, whether "Enforce constant definitions (enables some inlining)" is on or off, the error is the same. In DrRacket CS, when "Enforce constant definitions (enables some inlining)" is off, the error is different, blaming the definition of `x` instead.
3. In Racket CS, the error message uses "module: ..." whereas in Racket BC, the error message uses "in module:"

**Comment by sorawee** [comment]
Oh, interesting. For "../../racket/racket/share/pkgs/drracket/drracket/private/stack-checkpoint.rkt:115:30" to appear, "No debugging or profiling" must also be chosen.

**Comment by sorawee** [comment]
Another problem:

In Racket BC:

```
> (values (values 1 2))
; result arity mismatch;
;  expected number of values not received
;   expected: 1
;   received: 2
```

In Racket CS

```
> (values (values 1 2))
; returned two values to single value return context
```

**Comment by sorawee** [comment]
Similarly:

In Racket BC:

```
> (define-values (x y) 1)
; define-values: result arity mismatch;
;  expected number of values not received
;   expected: 2
;   received: 1
;   in: definition of x ...
```

In Racket CS:

```
> (define-values (x y) 1)
; define-values: incorrect number of values from rhs 1 (no newline)
```

**Comment by sorawee** [comment]
Another problem (sorry for spamming...):

Racket BC:

```
> (let-values ([(a b) (values 1 2 3)]) 1)
; result arity mismatch;
;  expected number of values not received
;   expected: 2
;   received: 3
;   in: local-binding form
```

Racket CS:

```
> (let-values ([(a b) (values 1 2 3)]) 1)
; result arity mismatch;
;  expected number of values not received
;   expected: 2
;   received: 3
;   at: local-binding form
```

Notice `in` vs `at`.

**Comment by gus-massa** [comment]
The last one looks easy to solve.  https://github.com/racket/racket/blob/454a586a79db4fef0492e2e70a532d05ca07ad1d/racket/src/cs/rumble/error.ss#L431

Perhaps you want to write the PR? (Assumin @mflatt doesn't disagree.)

**Comment by capfredf** [comment]
> Similarly:
> 
> In Racket BC:
> 
> ```
> > (define-values (x y) 1)
> ; define-values: result arity mismatch;
> ;  expected number of values not received
> ;   expected: 2
> ;   received: 1
> ;   in: definition of x ...
> ```
> 
> In Racket CS:
> 
> ```
> > (define-values (x y) 1)
> ; define-values: incorrect number of values from rhs 1 (no newline)
> ```

This difference doesn't appear if we put the definition in a program and run it.

I am guessing this is because at REPL, `raise-binding-result-arity-error` was called instead of `raise-definition-result-arity-error`, but I have some trouble figuring out which occurrence of raise-binding-result-arity-error should be replaced.  [ref](https://github.com/racket/racket/commit/ec9c30d2f158700460a1f7902ae9f6a52c150898)

@mflatt, can you give me some pointers on this issue?

**Comment by sorawee** [comment]
Hmm. It's not only that the error message is wrong. Programs that should error might not error at all!

```
#lang racket

y
(define y 1)
```

produces a value `'undefined5` when "Enforce constant definitions (enables some inlining)" is off.

We can even use this value in computations:

```
#lang racket

(eq? y y)
(define y 1)
```

produces `#t`.

**Comment by sorawee** [comment]
Just got hit by this again.

@mflatt I saw you changed several tests to accept error messages like "returned two values to single value return context" in addition to the one raised by `raise-result-arity-error`. Do I understand correctly that the difference is considered acceptable, and the bug is a WONTFIX?

**Comment by mflatt** [comment]
No, I just haven't gotten back to this.

**Comment by sorawee** [comment]
Here's another problem (which might be related to the above reported problem).

Consider an example in the [doc of `prop:exn:srclocs`](https://docs.racket-lang.org/reference/exns.html#%28def._%28%28quote._~23~25kernel%29._prop~3aexn~3asrclocs%29%29).

In DrRacket BC, DrRacket can highlight `voldemort` regardless of the setting of errortrace. The error message is:

```
oh dear don't say his name
```

In DrRacket CS, DrRacket can highlight `voldemort` only when errortrace is enabled. When it's disabled, `voldemort` is not highlighted, and the error message is:

```
 unsaved editor:42:17: oh dear don't say his name
```

**Comment by sorawee** [comment]
The remaining issues are also reported elsewhere (see above references), so let me close this.

**Comment by sorawee** [comment]
Let me also reopen this issue. There is one remaining issue left, which I will restate here:

```
#lang racket

(/ 1 0)
```

results in:

```
/: division by zero
```

in DrRacketBC, but results in:

```
projects/racket/extra-pkgs/drracket/drracket/drracket/private/stack-checkpoint.rkt:115:30: /: division by zero
```

in DrRacketCS.

Note that errortrace must be set to off.

I previously thought this bug would be fixed by the fix of racket/drracket#455, but they seems to be unrelated after all.

**Comment by sorawee** [comment]
The remaining issues are a combination of several bugs in DrRacket. 

- The `voldemort` problem is specific to only unsaved editor, and it's fixed by https://github.com/racket/drracket/commit/7c30a885a3ace297ffb15d7d73cf8fa3f82409e5 and https://github.com/racket/gui/commit/04e8164a122990606654ea00e7d01a58162ff6d7
- The `(/ 1 0)` problem is due to incorrect stack cutting, which is fixed in https://github.com/racket/drracket/pull/459. (The problem occurs in Racket BC too, but it's "hidden" because `continuation-mark-set->context` tends to give more information in Racket BC, which happens to hide the problem).

These are all fixed, so I will close the issue.
